### PR TITLE
[IOPLT-309] Adding Methods for reading from event hub

### PR DESCRIPTION
**Pull request by drmarro**
- Added new methods for reading from Event Hub using the fp-ts library.

#### List of Changes

Here's the files added and modified:
- KafkaConsumerCompact.ts: includes the consume features 

#### Motivation and Context

This change is necessary to enhance the handling of reading from Event Hub in our project.

#### How Has This Been Tested?

Unit tests are provided with the code.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: src/lib/KafkaConsumerCompact.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+): ConsumerRunConfig =>
+  ({
+    autoCommit: consumerRunOptions.autoCommit,
+    autoCommitInterval: consumerRunOptions.autoCommitInterval,
+    autoCommitThreshold: consumerRunOptions.autoCommitThreshold,
+    eachBatchAutoResolve: consumerRunOptions.eachBatchAutoResolve,
+    partitionsConsumedConcurrently:
+      consumerRunOptions.partitionsConsumedConcurrently,
+    eachMessage: eachMessageHandler
+  } as ConsumerRunConfig);
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:81]: We can safely remove casting here

File: src/lib/KafkaConsumerCompact.ts [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+): ConsumerRunConfig =>
+  ({
+    autoCommit: consumerRunOptions.autoCommit,
+    autoCommitInterval: consumerRunOptions.autoCommitInterval,
+    autoCommitThreshold: consumerRunOptions.autoCommitThreshold,
+    eachBatchAutoResolve: consumerRunOptions.eachBatchAutoResolve,
+    partitionsConsumedConcurrently:
+      consumerRunOptions.partitionsConsumedConcurrently,
+    eachBatch: eachBatchHandler
+  } as ConsumerRunConfig);
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:95]: ditto

File: src/lib/KafkaConsumerCompact.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+export enum ReadType {
+  Batch,
+  Message
+}
+
+export const read = (fa: KafkaConsumerCompact) => (
+  topic: string,
+  readType: ReadType = ReadType.Message,
+  handler?: (payload: KafkaConsumerPayload) => Promise<void>,
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:110]: why optional?

File: src/lib/KafkaConsumerCompact.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    {
+      brokers: [`${sas.url}:9093`],
+      ssl: true,
+      sasl: {
+        mechanism: "plain" as const,
+        username: "$ConnectionString",
+        password: AzureEventhubSasFromString.encode(sas)
+      },
+      clientId: sas.policy,
+      groupId: groupId || "consumer-group",
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:51]: can we use Option `fromNullable` here?

File: src/lib/KafkaConsumerCompact.ts [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+    eachBatchAutoResolve: consumerRunOptions.eachBatchAutoResolve,
+    partitionsConsumedConcurrently:
+      consumerRunOptions.partitionsConsumedConcurrently,
+    eachBatch: eachBatchHandler
+  } as ConsumerRunConfig);
+
+export const run = (config: ConsumerRunConfig) => (
+  client: Consumer
+): TE.TaskEither<Error, void> =>
+  TE.tryCatch(() => client.run(config), E.toError);
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:100]: What's the purpose of this function? Is it a single or a chunk message single operation? is it a wrapper form a consumer listener?

File: src/lib/KafkaConsumerCompact.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        autoCommitInterval: runnerConfig.autoCommitInterval,
+        autoCommitThreshold: runnerConfig.autoCommitThreshold,
+        eachBatchAutoResolve: runnerConfig.eachBatchAutoResolve,
+        partitionsConsumedConcurrently:
+          runnerConfig.partitionsConsumedConcurrently,
+        eachMessage: runnerConfig.handler,
+        eachBatch: undefined
+      })
+    )
+  );
```
Comment (AleDore) [inline src/lib/KafkaConsumerCompact.ts:103]:
```suggestion
pipe(
    {
      autoCommit: runnerConfig.autoCommit,
      autoCommitInterval: runnerConfig.autoCommitInterval,
      autoCommitThreshold: runnerConfig.autoCommitThreshold,
      eachBatchAutoResolve: runnerConfig.eachBatchAutoResolve,
      partitionsConsumedConcurrently:
        runnerConfig.partitionsConsumedConcurrently
    },
    commonConfig =>
      pipe(
        runnerConfig.readType === ReadType.Message,
        B.fold(
          () => ({
            ...commonConfig,
            eachBatch: runnerConfig.handler
          }),
          () => ({
            ...commonConfig,
            eachMessage: runnerConfig.handler,
            eachBatch: undefined
          })
        )
      )
  );
```

File: src/lib/KafkaTypes.ts [hunk 7]
```diff
@@ -10,3 +19,5 @@ export type KafkaProducerTopicConfig<T> = Omit<ProducerRecord, "messages"> & {
 
 export type KafkaProducerCompactConfig<T> = ValidableKafkaProducerConfig &
   KafkaProducerTopicConfig<T>;
+
+export type KafkaConsumerPayload = EachMessagePayload | EachBatchPayload;
```
Comment (AleDore) [inline src/lib/KafkaTypes.ts:23]: Do we need `EachMessagePayload ` ?

Comment (drmarro) [reply src/lib/KafkaTypes.ts]: This variable is not used anymore, I'll remove it.

**Review by AleDore — APPROVED** [review]

**Pull request merged**